sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadIntrinsics.java

```diff
@@ -0,0 +1,847 @@
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMLoadNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMStoreNode;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMNativePointer;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMPointer;
+import com.oracle.truffle.llvm.runtime.types.FunctionType;
+import com.oracle.truffle.llvm.runtime.types.PointerType;
+import com.oracle.truffle.llvm.runtime.types.Type;
+
+public class LLVMPThreadIntrinsics {
+    public static PrintWriter debugOut = null;
+
+    @CompilerDirectives.TruffleBoundary
+    private static void printDebug(String str) {
+        // create debug output tool
+        if (debugOut == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            try {
+                debugOut = new PrintWriter(new FileWriter("/home/florian/debug out"));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+        debugOut.println(str);
+        debugOut.flush();
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadCreate extends LLVMBuiltin {
+        @Child
+        LLVMStoreNode store = null;
+
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object thread, Object attr, Object startRoutine, Object arg) {
+            // create store node
+            if (store == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                store = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createStoreNode(LLVMInteropType.ValueKind.I64);
+            }
+
+            // create thread for execution of function
+            Thread t = LLVMLanguage.getLLVMContextReference().get().getEnv().createThread(() -> {
+                CompilerDirectives.transferToInterpreter();
+                // RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(getLLVMLanguage()));
+                RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(LLVMLanguage.getLanguage()));
+
+                // pthread_exit throws a control flow exception to stop the thread
+                try {
+                    callTarget.call(startRoutine, arg);
+                } catch (ControlFlowException e) {
+
+                }
+            });
+            // store cur id in thread var
+            store.executeWithTarget(thread, t.getId());
+
+            // store thread with thread id in context
+            LLVMLanguage.getLLVMContextReference().get().threadStorage.put(t.getId(), t);
+            // start thread
+            t.start();
+
+            return 0;
+        }
+    }
+
+    static final class MyArgNode extends LLVMExpressionNode {
+        private FrameSlot slot;
```

Comment:
This field should be final so partial evaluation can just remove the field access.